code/datums/components/spawner.dm

```diff
@@ -0,0 +1,70 @@
+/datum/component/spawner
+	///List of things we want to spawn at some point
+	var/mob_types = list(/mob/living/simple_animal/hostile/carp)
+	///How long we want it to be between spawns
+	var/spawn_time = 300 //30 seconds default
+	///Keeps track of all our mobs
+	var/list/spawned_mobs = list()
+	///delay till a new mob spawns
+	var/spawn_delay = 0
+	///Cap to prevent a massive pile of mobs
+	var/max_mobs = 5
+	///In case players can see the spawning mob, say if we stick the component on a xeno
+	var/spawn_text = "emerges from"
+	///Whether all the contents of mob_types should be spawned at once or not
+	var/squad_spawn = FALSE
+	///What faction we want this mob to be a part of
+	var/list/faction = list("xenomorph")
+
+
+/datum/component/spawner/Initialize(_mob_types, _spawn_time, _spawn_text, _max_mobs, _squad_spawn, _faction)
+	. = ..()	//Just in case
+	if(_spawn_time)
+		spawn_time=_spawn_time
+	if(_mob_types)
+		mob_types=_mob_types
+	if(_spawn_text)
+		spawn_text=_spawn_text
+	if(_max_mobs)
+		max_mobs=_max_mobs
+	if(_squad_spawn)
+		squad_spawn = _squad_spawn
+	if(_faction)
+		faction=_faction
+
+	RegisterSignal(parent, list(COMSIG_PARENT_QDELETING), .proc/stop_spawning)
+	START_PROCESSING(SSprocessing, src)
+
+/datum/component/spawner/process()
+	try_spawn_mob()
+
+
+/datum/component/spawner/proc/stop_spawning(force)
+	STOP_PROCESSING(SSprocessing, src)
+	spawned_mobs = null
+
+///Pro that checks whether we can spawn a mob and then spawns the mob in the specified fashion 
+/datum/component/spawner/proc/try_spawn_mob()

```

Comment:
use a subsystem for this please rather than ending up with 50+ copies of this component processing when an admin keeps making spawn effects because "xenos arent spawning so i just tried making more"